src/coreclr/jit/optimizer.cpp

```diff
@@ -8898,6 +8898,581 @@ bool Compiler::optIsRangeCheckRemovable(GenTree* tree)
+    {
+        return;
+    }
+
+    if (!varTypeIsIntegral(t2->TypeGet()) || !varTypeIsIntegral(t3->TypeGet()))
+    {
+        return;
+    }
+
+    /* The third block is Return with "CNS_INT int 0/1" */
+    if (t3->AsOp()->gtOp1->gtOper != GT_CNS_INT)
+    {
+        return;
+    }
+
+    if (t3->AsOp()->gtOp1->gtType != TYP_INT)
+    {
+        return;
+    }
+
+    /* Find the condition for the first block */
+
+    Statement* s1 = b1->lastStmt();
+
+    GenTree* t1 = s1->GetRootNode();
+    noway_assert(t1->gtOper == GT_JTRUE);
+
+    if (b2->countOfInEdges() > 1 || b3->countOfInEdges() > 1)
+    {
+        return;
+    }
+
+    /* Find the branch conditions of b1 and b2 */
+
+    bool bool1, bool2;
+
+    GenTree* c1 = optIsBoolComp(t1, &t1, &bool1);
+    if (!c1)
+    {
+        return;
+    }
+
+    GenTree* c2 = optIsBoolComp(t2, &t2, &bool2);
+    if (!c2)
+    {
+        return;
+    }
+
+    noway_assert(t1->OperIs(GT_EQ, GT_NE) && t1->AsOp()->gtOp1 == c1);
+    noway_assert(t2->OperIs(GT_EQ, GT_NE) && t2->AsOp()->gtOp1 == c2);
+
+    // Leave out floats where the bit-representation is more complicated
+    // - there are two representations for 0.
+    //
+    if (varTypeIsFloating(c1->TypeGet()) || varTypeIsFloating(c2->TypeGet()))
+    {
+        return;
+    }
+
+    // Make sure the types involved are of the same sizes
+    if (genTypeSize(c1->TypeGet()) != genTypeSize(c2->TypeGet()))
+    {
+        return;
+    }
+    if (genTypeSize(t1->TypeGet()) != genTypeSize(t2->TypeGet()))
+    {
+        return;
+    }
+#ifdef TARGET_ARMARCH
+    // Skip the small operand which we cannot encode.
+    if (varTypeIsSmall(c1->TypeGet()))
+        return;
+#endif
+    /* The second condition must not contain side effects */
+
+    if (c2->gtFlags & GTF_GLOB_EFFECT)
+    {
+        return;
+    }
+
+    /* The second condition must not be too expensive */
+
+    gtPrepareCost(c2);
+
+    if (c2->GetCostEx() > 12)
+    {
+        return;
+    }
+
+    genTreeOps foldOp;
+    genTreeOps cmpOp;
+    var_types  foldType = c1->TypeGet();
+    if (varTypeIsGC(foldType))
+    {
+        foldType = TYP_I_IMPL;
+    }
+
+    /* The b1 condition must be the reverse of the b2 condition */
+
+    if (t1->gtOper == t2->gtOper)
+    {
+        return;
+    }
+
+    ssize_t it3val = t3->AsOp()->gtOp1->AsIntCon()->gtIconVal;
+
+    if (t1->gtOper == GT_EQ && it3val == 0)
+    {
+        /* t1:c1==0 t2:c2!=0 t3:c3==0 ==> Return true if (c1&c2)!=0 */
+        foldOp = GT_AND;
+        cmpOp  = GT_NE;
+    }
+    else if (t1->gtOper == GT_NE && it3val == 0)
+    {
+        /* t1:c1!=0 t2:c2==0 t3:c3==0 ==> Return true if (c1|c2)==0 */
+        foldOp = GT_OR;
+        cmpOp  = GT_EQ;
+    }
+    else
+    {
+        /* Both cases requires NOT operation for operand. Do Not fold.
+             t1:c1==0 t2:c2!=0 t3:c3==1 ==> true if (c1&!c2)==0
+             t1:c1!=0 t2:c2==0 t3:c3==1 ==> true if (!c1&c2)==0 */
+        return;
+    }
+
```

Comment:
Seems like you could also handle these cases?
* `c1 == 0` and `c2 == 0` and `c3 == 1`  ==> `(c1 | c2) != 0`
* `c1 != 0` and `c2 != 0` and `c3 == 1` ==> `(c1 & c2) == 0`